### ISSUE TITLE
How to read a file from disk?

### ISSUE THREAD

**Issue by gq-vinted** [socketry/async-io#37]
It seems like most of the IO documentation focuses on network IO, so this is the code I tried:
```
def do_some_reading(task)
  task.async do |subtask|
    file = File.open('random_data.lol')
    generic = Async::IO::Generic.new(file)
    stream = Async::IO::Stream.new(generic)
    stream.read
  ensure
    file.close
  end
end
```
Running this multiple times in separate tasks with some debug output it looks like the tasks are being executed one after another, as if the IO code was blocking. This is the code I use to run it:
```
Async do |task|
  task.async { puts 'dummy task A' }

  lots_of_work.times do |n|
    task.async do |subtask|
      start_time = time - time0
      do_some_reading(subtask)
      end_time = time - time0

      puts "#{start_time},start"
      puts "#{end_time},end"
    end
  end

  task.async { puts 'dummy task B' }
end
```
And the output I get is:
```
dummy task A
0.0004913130542263389,start
1.807430032058619,end
1.8076503810007125,start
3.6459162089740857,end
3.64612788206432,start
5.541006584069692,end
5.541153504978865,start
7.373363309074193,end
7.373511067009531,start
9.267772497027181,end
dummy task B
```
What am I doing wrong? I realise reading from a disk drive won't be faster if I run it multiple times concurrently (this code is for a demonstration), but I'd expect both the dummy tasks to run almost immediately.

Also, using network IO gives me the results I expect - the tasks are executed concurrently and the first output I get is from both of the dummy tasks, so I think the issue I'm having is specifically with file IO.

**Comment by ioquatix** [comment]
We don't support non-blocking file I/O yet but it will come in the future with `io_uring`. As it stands, we could probably add a yield statement when doing buffered I/O on files so that we don't hog the even loop, i.e. insert `subtask.yield` after the file I/O to allow another task to execute.

Unfortunately we are beholden to the available system interfaces for non-blocking I/O.

**Comment by gq-vinted** [comment]
Gotcha, thank you for the answer. I'd assumed that async file I/O is a given and would be supported from the OS side, but it does seem like that does not exist (or isn't very portable). Interesting!

Looking at other approaches - is there an easy way to spawn a thread and wrap it up into a `Task` to wait for its result ansynchronously (thinking of something like an async `Thread#value`)? I'm not too experienced with Ruby, so not sure how much of a performance issue that might be. Would it be better to keep a thread pool for file I/O instead? Is this something I could do with one of the `async-*` libraries or would I have to build it from scratch? I appreciate these might be difficult to answer in the general case, but any hints would be welcome :).

Either way, nice to see `async` stuff in Ruby, thank you very much for putting in the work!

**Comment by ioquatix** [comment]
You can spin up a thread to write into a pipe and read the other end in async, if that's what you are after.

```ruby
#!/usr/bin/env ruby

require 'async'
require 'async/io'

output, input = IO.pipe

Thread.new do
	file = File.open("/dev/random", "r")
	while chunk = file.read(1024*8)
		puts "Writing chunk to pipe buffer..."
		input.write(chunk)
	end
end

Async do |task|
	output = Async::IO.try_convert(output)
	
	while chunk = output.read(1024)
		puts "#{chunk.bytesize}B read from pipe."
		task.sleep(1)
	end
end
```

**Comment by ioquatix** [comment]
> Looking at other approaches - is there an easy way to spawn a thread and wrap it up into a Task to wait for its result ansynchronously (thinking of something like an async Thread#value)?

Example above.

> I'm not too experienced with Ruby, so not sure how much of a performance issue that might be. 

Negligible unless you are literally processing 100k files.

> Would it be better to keep a thread pool for file I/O instead?

Ruby already has a thread pool internally, just use `Thread.new`.

> Is this something I could do with one of the async-* libraries or would I have to build it from scratch?

I probably need to understand what problem you are trying to solve before I can advise an existing solution.

> I appreciate these might be difficult to answer in the general case, but any hints would be welcome :).

No problem.

**Comment by gq-vinted** [comment]
Awesome. At the moment it's just a demo of how to use processes, threads and async in Ruby and how they handle different workloads, so for now I think I'll just go with `IO.pipe` in a thread. Again, thank you!

**Comment by ioquatix** [comment]
I flipped around input/output in the example to make more sense.

**Comment by ioquatix** [comment]
Have you tried doing something like this:

```
#!/usr/bin/env ruby

require 'async'
require 'async/io/stream'

Async do |task|
	file = File.open("/dev/random", "r")
	stream = Async::IO::Stream.new(file)
	
	while chunk = stream.read(1024)
		puts "#{chunk.bytesize}B read from pipe."
		task.yield
	end
end
```

Waiting on a mutex is not possible yet but we are discussing it. It needs additional support from the interpreter.

Using a pipe for synchronisation. It's even what `make` does.
